### PR TITLE
Error on invalid hex escape sequences #22

### PR DESCRIPTION
**Pull request by AngelJA**
Resolves: #22

- add new error: type error_invalid_hex_escape_sequence
- span returned by the error always points to the escape prefix `\x`

### CODE REVIEW

File: src/lex.cpp [hunk 1]
```diff
@@ -606,6 +606,10 @@ const char8* lexer::parse_string_literal() noexcept {
           ++c;
           break;
         }
+      case 'x':
+        // TODO (angel): do stuff here to identify invalid hex escape sequences?
+        // use parse_hexadecimal_number?
```
Comment (AngelJA) [inline src/lex.cpp:611]: looked at parse_hexadecimal_number more closely and this doesn't seem likely, just wanted to avoid writing redundant code

**Review by strager — CHANGES_REQUESTED** [review]
Discussed in person. Your test looks very wrong. The implementation looks like a good starting point.

**Comment by AngelJA** [comment]
I think I've fixed my test. I lengthened it a bit to include a valid and some invalid hex escapes. I left a todo to add more tests. I'm not great at test writing, so I'll spend some time studying yours to make sure I have good coverage.

Also added a simple implementation.

I've written this to output something like node would when trying to execute, "invalid hex escape sequence". But I've noticed other linters points to a single character with something like "expected hexadecimal digit". Preference?

File: test/test-lex.cpp [hunk 2]
```diff
@@ -649,10 +649,29 @@ TEST_F(test_lex, lex_strings) {
                                 offsets_matcher(input, 0, 14))));
       });
 
-  // TODO(strager): Report invalid hex escape sequences. For example:
```
Comment (strager) [inline test/test-lex.cpp:652]: Also include a positive test case (around line 550).

File: src/lex.cpp [hunk 3]
```diff
@@ -606,6 +606,16 @@ const char8* lexer::parse_string_literal() noexcept {
           ++c;
           break;
         }
+      case 'x':
+        for (int i = 0; i < 2; ++i) {
+          ++c;
+          if (!is_hex_digit(*c)) {
+            this->error_reporter_->report(error_invalid_hex_escape_sequence{
+                source_code_span(c-2-i, c+1)});
```
Comment (strager) [inline src/lex.cpp:614]: Instead of computing `c-2-i`, store the start of the escape sequence in a variable outside the loop (e.g. after line 609).

File: test/test-lex.cpp [hunk 4]
```diff
@@ -649,10 +649,29 @@ TEST_F(test_lex, lex_strings) {
...
-  //
-  // "hello\x1qworld"
-  // '\x'
+  // TODO (angel): add more tests
```
Comment (strager) [inline test/test-lex.cpp:652]: Some edge cases I thought of:
```javascript
"\x
```
```javascript
"\x1
```
```javascript
"\x"
```
```javascript
"\x\xyz"
```

File: src/lex.cpp (hunk 3 above)
Comment (AngelJA) [reply src/lex.cpp:614]: done, it made sense to me to place it slightly further up, let me know if that's no good

File: test/test-lex.cpp (hunk 2 above)
Comment (AngelJA) [reply test/test-lex.cpp:652]: Added two strings with hex escapes and no errors, not sure if you had something else in mind.

File: test/test-lex.cpp (hunk 4 above)
Comment (AngelJA) [reply test/test-lex.cpp:652]: Added all these test cases.

**Comment by AngelJA** [comment]
Made the changes mentioned  and left some comments.

For reference this test file: 
```javascript
let a = '\x;
let b = '\x1;
let c = '\x';
let d = '\x\xyz';
```
outputs this
```bash
test.js:1:10: error: invalid hex escape sequence: \x; [E060]
test.js:1:9: error: unclosed string literal [E040]
test.js:2:10: error: unexpected '\' in identifier [E043]
test.js:2:10: error: missing semicolon after expression [E027]
test.js:2:10: error: unexpected '\' in identifier [E043]
test.js:3:10: error: invalid hex escape sequence: \x' [E060]
test.js:4:10: error: invalid hex escape sequence: \x\ [E060]
test.js:4:12: error: invalid hex escape sequence: \xy [E060]
test.js:2:10: warning: use of undeclared variable: \x1 [E057]
```

File: test/test-lex.cpp [hunk 5]
```diff
@@ -649,10 +651,77 @@ TEST_F(test_lex, lex_strings) {
...
+      });
+
+  this->check_tokens_with_errors(
+      u8"'\\x1", {token_type::string},
+      [](padded_string_view input, const auto& errors) {
+        EXPECT_THAT(errors,
+                    ElementsAre(
+                      ERROR_TYPE_FIELD(
+                        error_invalid_hex_escape_sequence, escape_sequence,
+                        offsets_matcher(input, 1, 5)),
```
Comment (strager) [inline test/test-lex.cpp:676]: Blocking: 5 is out of bounds. The span should end at 4 instead.

Valid indexes are 0-3 (`'\x1`) and 4 (null terminator / end of file):
```
0: '
1: \
2: x
3: 1
4: (eof)
```

File: test/test-lex.cpp [hunk 6]
```diff
@@ -649,10 +651,77 @@ TEST_F(test_lex, lex_strings) {
...
+                        offsets_matcher(input, 0, 3))
+                    )
+        );
+      });
+
+  this->check_tokens_with_errors(
+      u8"'\\x1", {token_type::string},
+      [](padded_string_view input, const auto& errors) {
+        EXPECT_THAT(errors,
+                    ElementsAre(
```
Comment (strager) [inline test/test-lex.cpp:673]: Almost-blocking: The order of errors shouldn't matter. Use `UnorderedElementsAre` instead.

File: test/test-lex.cpp [hunk 7]
```diff
@@ -649,10 +651,77 @@ TEST_F(test_lex, lex_strings) {
...
+      });
+
+  this->check_tokens_with_errors(
+      u8"'\\x'", {token_type::string},
+      [](padded_string_view input, const auto& errors) {
+        EXPECT_THAT(errors,
+                    ElementsAre(
+                      ERROR_TYPE_FIELD(
+                        error_invalid_hex_escape_sequence, escape_sequence,
+                        offsets_matcher(input, 1, 4))
```
Comment (strager) [inline test/test-lex.cpp:691]: Minor: I'd expect the error squiggly to be on `\x` ([1-3)), not on `\x'` ([1-4)).

File: test/test-lex.cpp [hunk 8]
```diff
@@ -649,10 +651,77 @@ TEST_F(test_lex, lex_strings) {
...
+      u8"'\\x\\xyz'", {token_type::string},
+      [](padded_string_view input, const auto& errors) {
+        EXPECT_THAT(errors,
+                    ElementsAre(
+                      ERROR_TYPE_FIELD(
+                        error_invalid_hex_escape_sequence, escape_sequence,
+                        offsets_matcher(input, 1, 4)),
+                      ERROR_TYPE_FIELD(
+                        error_invalid_hex_escape_sequence, escape_sequence,
+                        offsets_matcher(input, 3, 6))
```
Comment (strager) [inline test/test-lex.cpp:706]: Hmm, these errors overlap. Is that a good idea? Not sure.

**Review by strager — APPROVED** [review]
Overall, looks good to me. Clean up the PR description and I'll merge it.

File: test/test-lex.cpp (hunk 6 above)
Comment (AngelJA) [reply test/test-lex.cpp:673]: Updated the tests where this applies.

File: test/test-lex.cpp (hunk 7 above)
Comment (AngelJA) [reply test/test-lex.cpp:691]: Updated to match this description. This also solves the out of bounds and overlap issues.

File: test/test-lex.cpp (hunk 5 above)
Comment (AngelJA) [reply test/test-lex.cpp]: This was resolved when the output was changed to only include `\x`.

File: test/test-lex.cpp (hunk 8 above)
Comment (AngelJA) [reply test/test-lex.cpp]: Some overlapping may be tough to avoid. For example, should invalid hex escape sequence also be raised inside an unclosed string?

Anyway, this error should never overlap anymore since it's been changed to always output `\x`.

**Comment by AngelJA** [comment]
The changes mentioned in the last review have all been made. Just lmk if you want me to also implement more specific errors, like incomplete hex escape, in addition to invalid hex escape.
